#Deep Neural Network for MNIST Classification Using TensorFlow

The dataset is called MNIST and refers to handwritten digit recognition. We can find more about it on Yann LeCun's website (Director of AI Research, Facebook). He is one of the pioneers of complex approaches that are widely used today, such as covolutional neural networks (CNNs). 

The dataset provides 70,000 images (28x28 pixels) of handwritten digits (1 digit per image). 

The goal is to write an algorithm that detects which digit is written. Since there are only 10 digits (0, 1, 2, 3, 4, 5, 6, 7, 8, 9), this is a classification problem with 10 classes.

In [0]:
# Importing Libraries

import numpy as np
import tensorflow as tf
import tensorflow_datasets as tfds

##Data

In [2]:
# Downloading the data from Tensorflow_datasets
mnist_dataset, mnist_info = tfds.load( name = "mnist", with_info="True", as_supervised="True")

# Split the training and testing datasets
mnist_train, mnist_test = mnist_dataset['train'], mnist_dataset['test']

# Assigning number of validation samples needed
num_validation_samples = 0.1*mnist_info.splits['train'].num_examples
num_validation_samples = tf.cast(num_validation_samples, tf.int64)

# Calculating number of test samples present
num_test_samples = mnist_info.splits['test'].num_examples
num_test_samples = tf.cast(num_test_samples, tf.int64)

# Defining a function which takes an image & label and returned scaled image & label
def scale(image, label):
  image = tf.cast(image, tf.float32)
  image /= 255
  return image, label

# Custom transform the training dataset using .map function
scaled_train_and_validation_data = mnist_train.map(scale)

# Similarly, transform the test dataset as well
test_data = mnist_test.map(scale)

# Shuffle the training dataset
BUFFER_SIZE = 10000
shuffled_train_and_validation_data = scaled_train_and_validation_data.shuffle(BUFFER_SIZE)

# Get the validation from the training data using .take method
validation_data = shuffled_train_and_validation_data.take(num_validation_samples)
train_data = shuffled_train_and_validation_data.skip(num_validation_samples)

BATCH_SIZE = 100
train_data = train_data.batch(BATCH_SIZE)
validation_data = validation_data.batch(num_validation_samples)

test_data = test_data.batch(num_test_samples)

validation_inputs, validation_targets = next(iter(validation_data))

local data directory. If you'd instead prefer to read directly from our public
GCS bucket (recommended if you're running on GCP), you can instead set
data_dir=gs://tfds-data/datasets.





Dataset mnist downloaded and prepared to /root/tensorflow_datasets/mnist/3.0.0. Subsequent calls will reuse this data.


##Model

In [0]:
input_size = 784
output_size = 10
hidden_layer_size = 50

model = tf.keras.Sequential([
                            tf.keras.layers.Flatten(input_shape=(28,28,1)),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(output_size, activation='softmax')
                            ])

###Choose the optimizer and loss function

In [0]:
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics= ['accuracy'])

###Training the built model

In [5]:
NUM_EPOCHS = 5

model.fit(train_data, epochs=NUM_EPOCHS, validation_data=(validation_inputs, validation_targets), verbose=2)

Epoch 1/5
540/540 - 10s - loss: 0.4085 - accuracy: 0.8848 - val_loss: 0.1963 - val_accuracy: 0.9455
Epoch 2/5
540/540 - 10s - loss: 0.1770 - accuracy: 0.9485 - val_loss: 0.1487 - val_accuracy: 0.9577
Epoch 3/5
540/540 - 11s - loss: 0.1353 - accuracy: 0.9605 - val_loss: 0.1237 - val_accuracy: 0.9628
Epoch 4/5
540/540 - 11s - loss: 0.1086 - accuracy: 0.9679 - val_loss: 0.1068 - val_accuracy: 0.9683
Epoch 5/5
540/540 - 11s - loss: 0.0923 - accuracy: 0.9727 - val_loss: 0.1021 - val_accuracy: 0.9688


###Testing the data

In [6]:
test_loss, test_accuracy = model.evaluate(test_data)

1/1 [==============================] - 0s 1ms/step - loss: 0.1064 - accuracy: 0.9675


In [7]:
print('Test loss: {0: .2f} & Test Accuracy: {1: .2f}%' .format(test_loss,test_accuracy*100))

Test loss:  0.11 & Test Accuracy:  96.75%


##Variations of the above model

In [8]:
# Downloading the data from Tensorflow_datasets
mnist_dataset, mnist_info = tfds.load( name = "mnist", with_info="True", as_supervised="True")

# Split the training and testing datasets
mnist_train, mnist_test = mnist_dataset['train'], mnist_dataset['test']

# Assigning number of validation samples needed
num_validation_samples = 0.1*mnist_info.splits['train'].num_examples
num_validation_samples = tf.cast(num_validation_samples, tf.int64)

# Calculating number of test samples present
num_test_samples = mnist_info.splits['test'].num_examples
num_test_samples = tf.cast(num_test_samples, tf.int64)

# Defining a function which takes an image & label and returned scaled image & label
def scale(image, label):
  image = tf.cast(image, tf.float32)
  image /= 255
  return image, label

# Custom transform the training dataset using .map function
scaled_train_and_validation_data = mnist_train.map(scale)

# Similarly, transform the test dataset as well
test_data = mnist_test.map(scale)

# Shuffle the training dataset
BUFFER_SIZE = 10000
shuffled_train_and_validation_data = scaled_train_and_validation_data.shuffle(BUFFER_SIZE)

# Get the validation from the training data using .take method
validation_data = shuffled_train_and_validation_data.take(num_validation_samples)
train_data = shuffled_train_and_validation_data.skip(num_validation_samples)

BATCH_SIZE = 50
train_data = train_data.batch(BATCH_SIZE)
validation_data = validation_data.batch(num_validation_samples)

test_data = test_data.batch(num_test_samples)

validation_inputs, validation_targets = next(iter(validation_data))


# Model

input_size = 784
output_size = 10
hidden_layer_size = 50

model = tf.keras.Sequential([
                            tf.keras.layers.Flatten(input_shape=(28,28,1)),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(output_size, activation='softmax')
                            ])

# Optimizer and loss function
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics= ['accuracy'])

# Training the model
NUM_EPOCHS = 5

model.fit(train_data, epochs=NUM_EPOCHS, validation_data=(validation_inputs, validation_targets), verbose=2)

# Testing the data
test_loss, test_accuracy = model.evaluate(test_data)
print('Test loss: {0: .2f} & Test Accuracy: {1: .2f}%' .format(test_loss,test_accuracy*100))

Epoch 1/5
1080/1080 - 12s - loss: 0.3377 - accuracy: 0.8995 - val_loss: 0.1755 - val_accuracy: 0.9468
Epoch 2/5
1080/1080 - 12s - loss: 0.1507 - accuracy: 0.9540 - val_loss: 0.1282 - val_accuracy: 0.9602
Epoch 3/5
1080/1080 - 12s - loss: 0.1142 - accuracy: 0.9655 - val_loss: 0.1148 - val_accuracy: 0.9653
Epoch 4/5
1080/1080 - 12s - loss: 0.0945 - accuracy: 0.9707 - val_loss: 0.0894 - val_accuracy: 0.9738
Epoch 5/5
1080/1080 - 12s - loss: 0.0791 - accuracy: 0.9753 - val_loss: 0.0845 - val_accuracy: 0.9747
1/1 [==============================] - 0s 1ms/step - loss: 0.1021 - accuracy: 0.9696
Test loss:  0.10 & Test Accuracy:  96.96%


In [9]:
# Downloading the data from Tensorflow_datasets
mnist_dataset, mnist_info = tfds.load( name = "mnist", with_info="True", as_supervised="True")

# Split the training and testing datasets
mnist_train, mnist_test = mnist_dataset['train'], mnist_dataset['test']

# Assigning number of validation samples needed
num_validation_samples = 0.1*mnist_info.splits['train'].num_examples
num_validation_samples = tf.cast(num_validation_samples, tf.int64)

# Calculating number of test samples present
num_test_samples = mnist_info.splits['test'].num_examples
num_test_samples = tf.cast(num_test_samples, tf.int64)

# Defining a function which takes an image & label and returned scaled image & label
def scale(image, label):
  image = tf.cast(image, tf.float32)
  image /= 255
  return image, label

# Custom transform the training dataset using .map function
scaled_train_and_validation_data = mnist_train.map(scale)

# Similarly, transform the test dataset as well
test_data = mnist_test.map(scale)

# Shuffle the training dataset
BUFFER_SIZE = 10000
shuffled_train_and_validation_data = scaled_train_and_validation_data.shuffle(BUFFER_SIZE)

# Get the validation from the training data using .take method
validation_data = shuffled_train_and_validation_data.take(num_validation_samples)
train_data = shuffled_train_and_validation_data.skip(num_validation_samples)

BATCH_SIZE = 1000
train_data = train_data.batch(BATCH_SIZE)
validation_data = validation_data.batch(num_validation_samples)

test_data = test_data.batch(num_test_samples)

validation_inputs, validation_targets = next(iter(validation_data))


# Model

input_size = 784
output_size = 10
hidden_layer_size = 100

model = tf.keras.Sequential([
                            tf.keras.layers.Flatten(input_shape=(28,28,1)),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                             tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                             tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(output_size, activation='softmax')
                            ])

# Optimizer and loss function
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics= ['accuracy'])

# Training the model
NUM_EPOCHS = 10

model.fit(train_data, epochs=NUM_EPOCHS, validation_data=(validation_inputs, validation_targets), verbose=2)

# Testing the data
test_loss, test_accuracy = model.evaluate(test_data)
print('Test loss: {0: .2f} & Test Accuracy: {1: .2f}%' .format(test_loss,test_accuracy*100))

Epoch 1/10
54/54 - 10s - loss: 0.9019 - accuracy: 0.7490 - val_loss: 0.3233 - val_accuracy: 0.9033
Epoch 2/10
54/54 - 10s - loss: 0.2577 - accuracy: 0.9244 - val_loss: 0.2105 - val_accuracy: 0.9373
Epoch 3/10
54/54 - 10s - loss: 0.1832 - accuracy: 0.9461 - val_loss: 0.1572 - val_accuracy: 0.9528
Epoch 4/10
54/54 - 10s - loss: 0.1421 - accuracy: 0.9574 - val_loss: 0.1278 - val_accuracy: 0.9620
Epoch 5/10
54/54 - 10s - loss: 0.1169 - accuracy: 0.9655 - val_loss: 0.1111 - val_accuracy: 0.9643
Epoch 6/10
54/54 - 10s - loss: 0.1018 - accuracy: 0.9691 - val_loss: 0.0957 - val_accuracy: 0.9705
Epoch 7/10
54/54 - 10s - loss: 0.0827 - accuracy: 0.9752 - val_loss: 0.0899 - val_accuracy: 0.9730
Epoch 8/10
54/54 - 10s - loss: 0.0756 - accuracy: 0.9775 - val_loss: 0.0814 - val_accuracy: 0.9748
Epoch 9/10
54/54 - 10s - loss: 0.0689 - accuracy: 0.9787 - val_loss: 0.0785 - val_accuracy: 0.9753
Epoch 10/10
54/54 - 10s - loss: 0.0616 - accuracy: 0.9813 - val_loss: 0.0620 - val_accuracy: 0.9812
1/1 [====

In [10]:
# Downloading the data from Tensorflow_datasets
mnist_dataset, mnist_info = tfds.load( name = "mnist", with_info="True", as_supervised="True")

# Split the training and testing datasets
mnist_train, mnist_test = mnist_dataset['train'], mnist_dataset['test']

# Assigning number of validation samples needed
num_validation_samples = 0.1*mnist_info.splits['train'].num_examples
num_validation_samples = tf.cast(num_validation_samples, tf.int64)

# Calculating number of test samples present
num_test_samples = mnist_info.splits['test'].num_examples
num_test_samples = tf.cast(num_test_samples, tf.int64)

# Defining a function which takes an image & label and returned scaled image & label
def scale(image, label):
  image = tf.cast(image, tf.float32)
  image /= 255
  return image, label

# Custom transform the training dataset using .map function
scaled_train_and_validation_data = mnist_train.map(scale)

# Similarly, transform the test dataset as well
test_data = mnist_test.map(scale)

# Shuffle the training dataset
BUFFER_SIZE = 10000
shuffled_train_and_validation_data = scaled_train_and_validation_data.shuffle(BUFFER_SIZE)

# Get the validation from the training data using .take method
validation_data = shuffled_train_and_validation_data.take(num_validation_samples)
train_data = shuffled_train_and_validation_data.skip(num_validation_samples)

BATCH_SIZE = 200
train_data = train_data.batch(BATCH_SIZE)
validation_data = validation_data.batch(num_validation_samples)

test_data = test_data.batch(num_test_samples)

validation_inputs, validation_targets = next(iter(validation_data))


# Model

input_size = 784
output_size = 10
hidden_layer_size = 50

model = tf.keras.Sequential([
                            tf.keras.layers.Flatten(input_shape=(28,28,1)),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                             tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                             tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(output_size, activation='softmax')
                            ])

# Optimizer and loss function
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics= ['accuracy'])

# Training the model
NUM_EPOCHS = 10

model.fit(train_data, epochs=NUM_EPOCHS, validation_data=(validation_inputs, validation_targets), verbose=2)

# Testing the data
test_loss, test_accuracy = model.evaluate(test_data)
print('Test loss: {0: .2f} & Test Accuracy: {1: .2f}%' .format(test_loss,test_accuracy*100))

Epoch 1/10
270/270 - 10s - loss: 0.5550 - accuracy: 0.8297 - val_loss: 0.2444 - val_accuracy: 0.9275
Epoch 2/10
270/270 - 10s - loss: 0.1973 - accuracy: 0.9407 - val_loss: 0.1621 - val_accuracy: 0.9503
Epoch 3/10
270/270 - 10s - loss: 0.1472 - accuracy: 0.9556 - val_loss: 0.1249 - val_accuracy: 0.9643
Epoch 4/10
270/270 - 10s - loss: 0.1231 - accuracy: 0.9629 - val_loss: 0.1116 - val_accuracy: 0.9663
Epoch 5/10
270/270 - 10s - loss: 0.1014 - accuracy: 0.9693 - val_loss: 0.1030 - val_accuracy: 0.9698
Epoch 6/10
270/270 - 10s - loss: 0.0881 - accuracy: 0.9732 - val_loss: 0.0834 - val_accuracy: 0.9758
Epoch 7/10
270/270 - 10s - loss: 0.0801 - accuracy: 0.9752 - val_loss: 0.0781 - val_accuracy: 0.9762
Epoch 8/10
270/270 - 10s - loss: 0.0692 - accuracy: 0.9789 - val_loss: 0.0742 - val_accuracy: 0.9783
Epoch 9/10
270/270 - 10s - loss: 0.0668 - accuracy: 0.9789 - val_loss: 0.0700 - val_accuracy: 0.9792
Epoch 10/10
270/270 - 10s - loss: 0.0546 - accuracy: 0.9828 - val_loss: 0.0705 - val_accura

In [11]:
# Downloading the data from Tensorflow_datasets
mnist_dataset, mnist_info = tfds.load( name = "mnist", with_info="True", as_supervised="True")

# Split the training and testing datasets
mnist_train, mnist_test = mnist_dataset['train'], mnist_dataset['test']

# Assigning number of validation samples needed
num_validation_samples = 0.1*mnist_info.splits['train'].num_examples
num_validation_samples = tf.cast(num_validation_samples, tf.int64)

# Calculating number of test samples present
num_test_samples = mnist_info.splits['test'].num_examples
num_test_samples = tf.cast(num_test_samples, tf.int64)

# Defining a function which takes an image & label and returned scaled image & label
def scale(image, label):
  image = tf.cast(image, tf.float32)
  image /= 255
  return image, label

# Custom transform the training dataset using .map function
scaled_train_and_validation_data = mnist_train.map(scale)

# Similarly, transform the test dataset as well
test_data = mnist_test.map(scale)

# Shuffle the training dataset
BUFFER_SIZE = 10000
shuffled_train_and_validation_data = scaled_train_and_validation_data.shuffle(BUFFER_SIZE)

# Get the validation from the training data using .take method
validation_data = shuffled_train_and_validation_data.take(num_validation_samples)
train_data = shuffled_train_and_validation_data.skip(num_validation_samples)

BATCH_SIZE = 100
train_data = train_data.batch(BATCH_SIZE)
validation_data = validation_data.batch(num_validation_samples)

test_data = test_data.batch(num_test_samples)

validation_inputs, validation_targets = next(iter(validation_data))


# Model

input_size = 784
output_size = 10
hidden_layer_size = 50

model = tf.keras.Sequential([
                            tf.keras.layers.Flatten(input_shape=(28,28,1)),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(output_size, activation='softmax')
                            ])

# Optimizer and loss function
custom_adam = tf.keras.optimizers.Adam(learning_rate=0.0001)
model.compile(optimizer= custom_adam, loss='sparse_categorical_crossentropy', metrics= ['accuracy'])

# Training the model
NUM_EPOCHS = 10

model.fit(train_data, epochs=NUM_EPOCHS, validation_data=(validation_inputs, validation_targets), verbose=2)

# Testing the data
test_loss, test_accuracy = model.evaluate(test_data)
print('Test loss: {0: .2f} & Test Accuracy: {1: .2f}%' .format(test_loss,test_accuracy*100))

Epoch 1/10
540/540 - 11s - loss: 1.2404 - accuracy: 0.6461 - val_loss: 0.4691 - val_accuracy: 0.8782
Epoch 2/10
540/540 - 11s - loss: 0.3851 - accuracy: 0.8942 - val_loss: 0.3155 - val_accuracy: 0.9122
Epoch 3/10
540/540 - 11s - loss: 0.2971 - accuracy: 0.9149 - val_loss: 0.2640 - val_accuracy: 0.9265
Epoch 4/10
540/540 - 11s - loss: 0.2596 - accuracy: 0.9253 - val_loss: 0.2341 - val_accuracy: 0.9353
Epoch 5/10
540/540 - 11s - loss: 0.2330 - accuracy: 0.9328 - val_loss: 0.2160 - val_accuracy: 0.9422
Epoch 6/10
540/540 - 11s - loss: 0.2145 - accuracy: 0.9383 - val_loss: 0.2034 - val_accuracy: 0.9432
Epoch 7/10
540/540 - 11s - loss: 0.1989 - accuracy: 0.9420 - val_loss: 0.1867 - val_accuracy: 0.9467
Epoch 8/10
540/540 - 11s - loss: 0.1867 - accuracy: 0.9457 - val_loss: 0.1743 - val_accuracy: 0.9512
Epoch 9/10
540/540 - 11s - loss: 0.1730 - accuracy: 0.9490 - val_loss: 0.1640 - val_accuracy: 0.9552
Epoch 10/10
540/540 - 11s - loss: 0.1632 - accuracy: 0.9524 - val_loss: 0.1562 - val_accura

In [12]:
# Downloading the data from Tensorflow_datasets
mnist_dataset, mnist_info = tfds.load( name = "mnist", with_info="True", as_supervised="True")

# Split the training and testing datasets
mnist_train, mnist_test = mnist_dataset['train'], mnist_dataset['test']

# Assigning number of validation samples needed
num_validation_samples = 0.1*mnist_info.splits['train'].num_examples
num_validation_samples = tf.cast(num_validation_samples, tf.int64)

# Calculating number of test samples present
num_test_samples = mnist_info.splits['test'].num_examples
num_test_samples = tf.cast(num_test_samples, tf.int64)

# Defining a function which takes an image & label and returned scaled image & label
def scale(image, label):
  image = tf.cast(image, tf.float32)
  image /= 255
  return image, label

# Custom transform the training dataset using .map function
scaled_train_and_validation_data = mnist_train.map(scale)

# Similarly, transform the test dataset as well
test_data = mnist_test.map(scale)

# Shuffle the training dataset
BUFFER_SIZE = 10000
shuffled_train_and_validation_data = scaled_train_and_validation_data.shuffle(BUFFER_SIZE)

# Get the validation from the training data using .take method
validation_data = shuffled_train_and_validation_data.take(num_validation_samples)
train_data = shuffled_train_and_validation_data.skip(num_validation_samples)

BATCH_SIZE = 100
train_data = train_data.batch(BATCH_SIZE)
validation_data = validation_data.batch(num_validation_samples)

test_data = test_data.batch(num_test_samples)

validation_inputs, validation_targets = next(iter(validation_data))


# Model

input_size = 784
output_size = 10
hidden_layer_size = 50

model = tf.keras.Sequential([
                            tf.keras.layers.Flatten(input_shape=(28,28,1)),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(output_size, activation='softmax')
                            ])

# Optimizer and loss function
custom_adam = tf.keras.optimizers.Adam(learning_rate=0.001)
model.compile(optimizer= custom_adam, loss='sparse_categorical_crossentropy', metrics= ['accuracy'])

# Training the model
NUM_EPOCHS = 20

model.fit(train_data, epochs=NUM_EPOCHS, validation_data=(validation_inputs, validation_targets), verbose=2)

# Testing the data
test_loss, test_accuracy = model.evaluate(test_data)
print('Test loss: {0: .2f} & Test Accuracy: {1: .2f}%' .format(test_loss,test_accuracy*100))

Epoch 1/20
540/540 - 11s - loss: 0.4377 - accuracy: 0.8641 - val_loss: 0.1879 - val_accuracy: 0.9463
Epoch 2/20
540/540 - 11s - loss: 0.1643 - accuracy: 0.9510 - val_loss: 0.1480 - val_accuracy: 0.9583
Epoch 3/20
540/540 - 11s - loss: 0.1273 - accuracy: 0.9614 - val_loss: 0.1132 - val_accuracy: 0.9668
Epoch 4/20
540/540 - 11s - loss: 0.1059 - accuracy: 0.9677 - val_loss: 0.1049 - val_accuracy: 0.9692
Epoch 5/20
540/540 - 11s - loss: 0.0906 - accuracy: 0.9722 - val_loss: 0.0938 - val_accuracy: 0.9730
Epoch 6/20
540/540 - 12s - loss: 0.0792 - accuracy: 0.9748 - val_loss: 0.0924 - val_accuracy: 0.9718
Epoch 7/20
540/540 - 11s - loss: 0.0712 - accuracy: 0.9776 - val_loss: 0.0801 - val_accuracy: 0.9782
Epoch 8/20
540/540 - 11s - loss: 0.0612 - accuracy: 0.9807 - val_loss: 0.0730 - val_accuracy: 0.9792
Epoch 9/20
540/540 - 11s - loss: 0.0576 - accuracy: 0.9824 - val_loss: 0.0720 - val_accuracy: 0.9805
Epoch 10/20
540/540 - 11s - loss: 0.0508 - accuracy: 0.9836 - val_loss: 0.0724 - val_accura

In [13]:
# Downloading the data from Tensorflow_datasets
mnist_dataset, mnist_info = tfds.load( name = "mnist", with_info="True", as_supervised="True")

# Split the training and testing datasets
mnist_train, mnist_test = mnist_dataset['train'], mnist_dataset['test']

# Assigning number of validation samples needed
num_validation_samples = 0.1*mnist_info.splits['train'].num_examples
num_validation_samples = tf.cast(num_validation_samples, tf.int64)

# Calculating number of test samples present
num_test_samples = mnist_info.splits['test'].num_examples
num_test_samples = tf.cast(num_test_samples, tf.int64)

# Defining a function which takes an image & label and returned scaled image & label
def scale(image, label):
  image = tf.cast(image, tf.float32)
  image /= 255
  return image, label

# Custom transform the training dataset using .map function
scaled_train_and_validation_data = mnist_train.map(scale)

# Similarly, transform the test dataset as well
test_data = mnist_test.map(scale)

# Shuffle the training dataset
BUFFER_SIZE = 10000
shuffled_train_and_validation_data = scaled_train_and_validation_data.shuffle(BUFFER_SIZE)

# Get the validation from the training data using .take method
validation_data = shuffled_train_and_validation_data.take(num_validation_samples)
train_data = shuffled_train_and_validation_data.skip(num_validation_samples)

BATCH_SIZE = 100
train_data = train_data.batch(BATCH_SIZE)
validation_data = validation_data.batch(num_validation_samples)

test_data = test_data.batch(num_test_samples)

validation_inputs, validation_targets = next(iter(validation_data))


# Model

input_size = 784
output_size = 10
hidden_layer_size = 50

model = tf.keras.Sequential([
                            tf.keras.layers.Flatten(input_shape=(28,28,1)),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
                            tf.keras.layers.Dense(output_size, activation='softmax')
                            ])

# Optimizer and loss function
custom_adam = tf.keras.optimizers.Adam(learning_rate=0.001)
model.compile(optimizer= custom_adam, loss='sparse_categorical_crossentropy', metrics= ['accuracy'])

# Training the model
NUM_EPOCHS = 30

model.fit(train_data, epochs=NUM_EPOCHS, validation_data=(validation_inputs, validation_targets), verbose=2)

# Testing the data
test_loss, test_accuracy = model.evaluate(test_data)
print('Test loss: {0: .2f} & Test Accuracy: {1: .2f}%' .format(test_loss,test_accuracy*100))

Epoch 1/30
540/540 - 11s - loss: 0.4113 - accuracy: 0.8790 - val_loss: 0.1986 - val_accuracy: 0.9403
Epoch 2/30
540/540 - 11s - loss: 0.1695 - accuracy: 0.9491 - val_loss: 0.1351 - val_accuracy: 0.9612
Epoch 3/30
540/540 - 12s - loss: 0.1261 - accuracy: 0.9620 - val_loss: 0.1283 - val_accuracy: 0.9622
Epoch 4/30
540/540 - 11s - loss: 0.1051 - accuracy: 0.9684 - val_loss: 0.0970 - val_accuracy: 0.9720
Epoch 5/30
540/540 - 11s - loss: 0.0841 - accuracy: 0.9740 - val_loss: 0.0881 - val_accuracy: 0.9763
Epoch 6/30
540/540 - 12s - loss: 0.0765 - accuracy: 0.9763 - val_loss: 0.0798 - val_accuracy: 0.9783
Epoch 7/30
540/540 - 11s - loss: 0.0666 - accuracy: 0.9793 - val_loss: 0.0675 - val_accuracy: 0.9807
Epoch 8/30
540/540 - 11s - loss: 0.0585 - accuracy: 0.9814 - val_loss: 0.0687 - val_accuracy: 0.9788
Epoch 9/30
540/540 - 11s - loss: 0.0506 - accuracy: 0.9841 - val_loss: 0.0618 - val_accuracy: 0.9812
Epoch 10/30
540/540 - 11s - loss: 0.0483 - accuracy: 0.9841 - val_loss: 0.0540 - val_accura

##Conclusion

No matter how much we changed the values of hyperparameters of the model, we got the test accuracy as close to 97%.